src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Clear.cs

```diff
@@ -0,0 +1,126 @@
+using Microsoft.PowerFx.Syntax;
+
+namespace Microsoft.PowerFx.Core.Texl.Builtins
+{
+    // Clear(collection:*[])
+    internal class ClearFunction : BuiltinFunction
+    {
+        public override bool ManipulatesCollections => true;
+
+        public override bool ModifiesValues => true;
+
+        public override bool IsSelfContained => false;
+
+        public override bool AllowedWithinNondeterministicOperationOrder => false;
+
+        public override bool SupportsParamCoercion => false;
+
+        public override bool MutatesArg0 => true;
+
+        public ClearFunction()
+            : base("Clear", TexlStrings.AboutClear, FunctionCategories.Behavior, DType.Unknown, 0, 1, 1, DType.EmptyTable)
+        {
+        }
+
+        public override IEnumerable<TexlStrings.StringGetter[]> GetSignatures()
+        {
+            yield return new[] { TexlStrings.ClearArg1 };
+        }
+
+        public override bool CheckTypes(CheckTypesContext context, TexlNode[] args, DType[] argTypes, IErrorContainer errors, out DType returnType, out Dictionary<TexlNode, DType> nodeToCoercedTypeMap)
+        {
+            Contracts.AssertValue(args);
+            Contracts.AssertAllValues(args);
+            Contracts.AssertValue(argTypes);
+            Contracts.Assert(args.Length == argTypes.Length);
+            Contracts.AssertValue(errors);
+            Contracts.Assert(MinArity <= args.Length && args.Length <= MaxArity);
+
+            bool fValid = base.CheckTypes(context, args, argTypes, errors, out returnType, out nodeToCoercedTypeMap);
+            Contracts.Assert(returnType.IsUnknown);
+
+            DType collectionType = argTypes[0];
+            if (!collectionType.IsTable)
+            {
+                fValid = false;
+                errors.EnsureError(args[0], TexlStrings.ErrNeedTable_Func, Name);
+            }
+
+            returnType = context.Features.PowerFxV1CompatibilityRules ? DType.Void : collectionType;
+
+            return fValid;
+        }
+
+        public override void CheckSemantics(TexlBinding binding, TexlNode[] args, DType[] argTypes, IErrorContainer errors)
+        {
+            Contracts.AssertValue(binding);
+            Contracts.AssertValue(args);
+            Contracts.AssertAllValues(args);
+            Contracts.AssertValue(argTypes);
+            Contracts.Assert(args.Length == argTypes.Length);
+            Contracts.AssertValue(errors);
+            Contracts.Assert(MinArity <= args.Length && args.Length <= MaxArity);
+
+            DType dataSourceType = argTypes[0];
+
+            if (binding.Features.PowerFxV1CompatibilityRules && dataSourceType.IsTable && dataSourceType.Kind != DKind.ObjNull)
+            {
+                base.ValidateArgumentIsMutable(binding, args[0], errors);
+            }
+
+            if (binding.EntityScope != null && 
+                binding.EntityScope.TryGetDataSource(args[0], out IExternalDataSource dataSourceInfo) &&
+                !dataSourceInfo.IsClearable)
```

Comment:
>          !dataSourceInfo.IsClearable) [](http://example.com/codeflow?start=6&length=38)

This is good - 
Today, we totally block Clear() in certain scenarios to prevent use on wrong sources; but this means we could allow the function, and the enforcement now happens here. 

That would let makers use Clear() on safe targets (like in-memory collections) while blocking on unsafe targets (like DV tables)